bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -0,0 +1,265 @@
+	/// The origin address
+	pub origin: H160,
+	/// The assets
+	pub assets: Vec<Asset>,
+	// The command originating from the Gateway contract
+	pub xcm: Vec<u8>,
+	// The claimer in the case that funds get trapped.
+	pub claimer: Option<Vec<u8>>,
+}
+
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum Asset {
+	NativeTokenERC20 {
+		/// The native token ID
+		token_id: H160,
+		/// The monetary value of the asset
+		value: u128,
+	},
+	ForeignTokenERC20 {
+		/// The foreign token ID
+		token_id: H256,
+		/// The monetary value of the asset
+		value: u128,
+	},
+}
+
+/// Reason why a message conversion failed.
+#[derive(Copy, Clone, TypeInfo, PalletError, Encode, Decode, RuntimeDebug)]
+pub enum ConvertMessageError {
+	/// The XCM provided with the message could not be decoded into XCM.
+	InvalidXCM,
+	/// The XCM provided with the message could not be decoded into versioned XCM.
+	InvalidVersionedXCM,
+	/// Invalid claimer MultiAddress provided in payload.
+	InvalidClaimer,
+	/// Invalid foreign ERC20 token ID
+	InvalidAsset,
+}
+
+pub trait ConvertMessage {
+	fn convert(message: Message, origin_account: Location) -> Result<Xcm<()>, ConvertMessageError>;
+}
+
+pub struct MessageToXcm<EthereumNetwork, InboundQueuePalletInstance, ConvertAssetId, XcmPrologueFee>
+where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	XcmPrologueFee: Get<u128>,
+{
+	_phantom:
+		PhantomData<(EthereumNetwork, InboundQueuePalletInstance, ConvertAssetId, XcmPrologueFee)>,
+}
+
+impl<EthereumNetwork, InboundQueuePalletInstance, ConvertAssetId, XcmPrologueFee> ConvertMessage
+	for MessageToXcm<EthereumNetwork, InboundQueuePalletInstance, ConvertAssetId, XcmPrologueFee>
+where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	XcmPrologueFee: Get<u128>,
+{
+	fn convert(
+		message: Message,
+		origin_account_location: Location,
+	) -> Result<Xcm<()>, ConvertMessageError> {
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			// Decode xcm
+			let versioned_xcm = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			)
+			.map_err(|_| ConvertMessageError::InvalidVersionedXCM)?;
+			message_xcm = versioned_xcm.try_into().map_err(|_| ConvertMessageError::InvalidXCM)?;
+		}
+
+		log::debug!(target: LOG_TARGET,"xcm decoded as {:?}", message_xcm);
+
+		let network = EthereumNetwork::get();
+
+		let origin_location = Location::new(
+			2,
+			[GlobalConsensus(network), AccountKey20 { key: message.origin.into(), network: None }],
+		);
+
+		let network = EthereumNetwork::get();
+
+		let fee_asset = Location::new(1, Here);
+		let fee: xcm::prelude::Asset = (fee_asset.clone(), XcmPrologueFee::get()).into();
+		let mut instructions = vec![
+			ReceiveTeleportedAsset(fee.clone().into()),
+			PayFees { asset: fee },
```

Comment:
Not sure this will work for advanced use cases like 2 hops transfer or transact.

e.g.  In V1 for register ERC20 on AH there is a `create_asset` fee required we need to prefund the bridge location which acts as the owner of the asset.

Reply:
Specifically, follow this in code in Contracts V2 for issuing token registration messages: https://github.com/Snowfork/snowbridge/blob/553b0b9d9ec6f3683c6c605417ea5532282771ed/contracts/src/v2/Calls.sol#L115

Reply:
Yeah on its my todo 👍

Reply:
@claravanstaden @vgeddes 
Since user already paid Ether on Ethereum, why not just use it as fee?  Then we don't need burn/teleport on BH from the relayer.

https://github.com/claravanstaden/polkadot-sdk/pull/5 for demonstration, please check if that make sense.

Reply:
> AH will just accept any sufficient foreign asset as fee

These are always exchanged to Dot behind the scenes automatically. If the weth/dot liquidity pool is empty/broken there will be a failure.

> Then we don't even need a separated round of the reward process

I prefer having a separate reward process. It has several benefits, like giving the relayer the option of choosing their account where claimed rewards can be deposited.

Reply:
> If the weth/dot liquidity pool is empty/broken there will be a failure.

Btw: weth/dot liquidity pool is not required on AH, there are multiple traders configured there.

https://github.com/paritytech/polkadot-sdk/blob/2ef2723126584dfcd6d2a9272282ee78375dbcd3/cumulus/parachains/runtimes/assets/asset-hub-westend/src/xcm_config.rs#L437-L449 use ED as exchange rate by default.